lib/vnet/setup.go

```diff
@@ -0,0 +1,130 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"log/slog"
+	"os"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	"golang.zx2c4.com/wireguard/tun"
+)
+
+// Run is a blocking call to create and start Teleport VNet.
+func Run(ctx context.Context) error {
+	ipv6Prefix, err := IPv6Prefix()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tun, err := CreateAndSetupTUNDevice(ctx, ipv6Prefix.String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	manager, err := NewManager(&Config{
+		TUNDevice:  tun,
+		IPv6Prefix: ipv6Prefix,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error { return trace.Wrap(manager.Run(ctx)) })
+	g.Go(func() error {
+		<-ctx.Done()
+		tunErr := tun.Close()
+		destroyErr := manager.Destroy()
+		return trace.NewAggregate(tunErr, destroyErr)
+	})
+	return trace.Wrap(g.Wait())
+}
+
+// AdminSubcommand is the tsh subcommand that should run as root that will
+// create and setup a TUN device and pass the file descriptor for that device
+// over the unix socket found at socketPath.
+func AdminSubcommand(ctx context.Context, socketPath, ipv6Prefix string) error {
+	tun, tunName, err := createAndSetupTUNDeviceAsRoot(ctx, ipv6Prefix)
+	if err != nil {
+		return trace.Wrap(err, "performing admin setup")
+	}
+	if err := sendTUNNameAndFd(socketPath, tunName, tun.File().Fd()); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// CreateAndSetupTUNDevice returns a virtual network device and configures the host OS to use that device for
+// VNet connections.
+func CreateAndSetupTUNDevice(ctx context.Context, ipv6Prefix string) (tun.Device, error) {
+	var (
+		device tun.Device
+		name   string
+		err    error
+	)
+	if os.Getuid() == 0 {
+		// We can get here if the user runs `tsh vnet` as root, but it is not in the expected path when
+		// started as a regular user, AdminSubcommand directly calls createAndSetupTUNDeviceAsRoot.
+		device, name, err = createAndSetupTUNDeviceAsRoot(ctx, ipv6Prefix)
+	} else {
+		device, name, err = createAndSetupTUNDeviceWithoutRoot(ctx, ipv6Prefix)
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	slog.InfoContext(ctx, "Created TUN device.", "device", name)
+	return device, nil
+}
+
+func createAndSetupTUNDeviceAsRoot(ctx context.Context, ipv6Prefix string) (tun.Device, string, error) {
+	tun, tunName, err := createTUNDevice(ctx)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	tunIPv6 := ipv6Prefix + "1"
```

Comment:
well, this part should be okay because the prefix has 40 random bits and is generated on each startup, but you still shouldn't run VNet concurrently on the same machine because DNS configuration and IPv4 ranges (once we add IPv4) are likely to conflict

Reply:
Ah right, fair point. Is that something we should check for (either in this PR or when we add IPv4)?

Reply:
added it to the roadmap